File: pkg/models/customer.go
package models

import "time"

// CustomerType represents customer type
type CustomerType int8

// Customer types
const (
	CUSTOMER_TYPE_CUSTOMER    CustomerType = 1 // 客户
	CUSTOMER_TYPE_SUPPLIER    CustomerType = 2 // 供应商
	CUSTOMER_TYPE_BOTH        CustomerType = 3 // 既是客户也是供应商
)

type Customer struct {
	Id  int64  `xorm:"PK"`
	Uid int64  `xorm:"INDEX(IDX_customer_uid_deleted) NOT NULL"`
	Deleted bool `xorm:"INDEX(IDX_customer_uid_deleted) NOT NULL"`
	// 客户名称
	Name string `xorm:"VARCHAR(256) NOT NULL"`
	// 客户类型 1:客户 2:供应商 3:既是客户也是供应商
	CustomerType int8 `xorm:"DEFAULT 1 NOT NULL"`
	// 地址
	Address string `xorm:"VARCHAR(512) DEFAULT ''"`
	// 联系人
	Contacts string `xorm:"VARCHAR(256) DEFAULT ''"`
	// 联系方式
	ContactsInfo string `xorm:"VARCHAR(256) DEFAULT ''"`
	// 备注
	Comment string `xorm:"VARCHAR(512) DEFAULT ''"`
	// 是否隐藏
	Hidden bool `xorm:"DEFAULT false NOT NULL"`
	// 创建/更新时间
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

// ToCustomerInfoResponse returns a view-object according to database model
func (c *Customer) ToCustomerInfoResponse(userTimezone *time.Location) *CustomerInfo {
	return &CustomerInfo{
		Id:           c.Id,
		Name:         c.Name,
		CustomerType: c.CustomerType,
		Address:      c.Address,
		Contacts:     c.Contacts,
		ContactsInfo: c.ContactsInfo,
		Comment:      c.Comment,
		Hidden:       c.Hidden,
		CreatedTime:  time.Unix(c.CreatedUnixTime, 0).In(userTimezone),
		UpdatedTime:  time.Unix(c.UpdatedUnixTime, 0).In(userTimezone),
	}
}

File: src/models/customer.ts
// Customer types
export enum CustomerType {
    CUSTOMER = 1,
    SUPPLIER = 2,
    BOTH = 3
}

export interface CustomerInfo {
    id: string;
    name: string;
    customer_type: CustomerType;
    address: string;
    contacts: string;
    contacts_info: string;
    comment: string;
    hidden: boolean;
    created_time: string;
    updated_time: string;
}

export interface CustomerListResponse {
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
    customers: CustomerInfo[];
}

export interface CustomerCreateRequest {
    client_session_id?: string;
    name: string;
    customer_type: CustomerType;
    address?: string;
    contacts?: string;
    contacts_info?: string;
    comment?: string;
    hidden?: boolean;
}

export interface CustomerModifyRequest {
    id: string;
    name: string;
    customer_type: CustomerType;
    address?: string;
    contacts?: string;
    contacts_info?: string;
    comment?: string;
    hidden?: boolean;
}

export interface CustomerDeleteRequest {
    id: string;
}

export interface CustomerHideRequest {
    id: string;
    hidden: boolean;
}

export interface CustomerListRequest {
    visible_only?: boolean;
    customer_type?: CustomerType;
    page?: number;
    page_size?: number;
}

export class Customer {
    public id: string;
    public name: string;
    public customerType: CustomerType;
    public address: string;
    public contacts: string;
    public contactsInfo: string;
    public comment: string;
    public hidden: boolean;
    public createdTime: Date;
    public updatedTime: Date;

    private constructor(info: CustomerInfo) {
        this.id = info.id;
        this.name = info.name;
        this.customerType = info.customer_type;
        this.address = info.address;
        this.contacts = info.contacts;
        this.contactsInfo = info.contacts_info;
        this.comment = info.comment;
        this.hidden = info.hidden;
        this.createdTime = new Date(info.created_time);
        this.updatedTime = new Date(info.updated_time);
    }

    public get visible(): boolean {
        return !this.hidden;
    }

    public get customerTypeName(): string {
        switch (this.customerType) {
            case CustomerType.CUSTOMER:
                return 'Customer';
            case CustomerType.SUPPLIER:
                return 'Supplier';
            case CustomerType.BOTH:
                return 'Both';
            default:
                return '';
        }
    }

    public equals(other: Customer): boolean {
        return this.id === other.id &&
            this.name === other.name &&
            this.customerType === other.customerType &&
            this.address === other.address &&
            this.contacts === other.contacts &&
            this.contactsInfo === other.contactsInfo &&
            this.comment === other.comment &&
            this.hidden === other.hidden;
    }

    public toModifyRequest(): CustomerModifyRequest {
        return {
            id: this.id,
            name: this.name,
            customer_type: this.customerType,
            address: this.address,
            contacts: this.contacts,
            contacts_info: this.contactsInfo,
            comment: this.comment,
            hidden: this.hidden
        };
    }

    public static of(info: CustomerInfo): Customer {
        return new Customer(info);
    }

    public static ofMulti(infos: CustomerInfo[]): Customer[] {
        return infos.map(info => Customer.of(info));
    }

    public static createNew(name: string, customerType: CustomerType): Customer {
        return new Customer({
            id: '',
            name: name,
            customer_type: customerType,
            address: '',
            contacts: '',
            contacts_info: '',
            comment: '',
            hidden: false,
            created_time: new Date().toISOString(),
            updated_time: new Date().toISOString()
        });
    }
}
